zebra: rework RA handling for vrf-lite

sworleys opened the pull request:
Rework RA handling for vrf-lite scenarios.

Before we were using a single FD descriptor for polling
across multiple zvrf's. This would cause us to hit this
assert() in some bgp unnumbered and vrrp configs:

```
/*
 * What happens if we have a thread already
 * created for this event?
 */
if (thread_array[fd])
	assert(!"Thread already scheduled for file descriptor");
```

We were scheduling a thread_read on the same FD for every zvrf.

With vrf-lite, RAs and ARPs are not vrf-bound, so we can just use one
rtadv instance to manage them for all VRFs. We will choose the default
VRF for this.

This patch removes the rtadv_sock altogether for zrouter and moves the
functionality this represented to the default VRF. All RAs will be
handled in the default VRF under vrf-lite configs with only one poll
thread started for it.

This patch also extends how we track subscribed interfaces (s or msec)
to use an actual sorted list by interface names rather than just a
counter. With multiple daemons turning interfaces/on/off these counters
can get very wrong during ifup/down events. Making them a sorted list
prevents this from happening by preventing duplicates.

With netns-vrf's nothing should change other than the interface list.

Signed-off-by: Stephen Worley <sworley@nvidia.com>

pguibert6WIND commented on zebra/rtadv.c:
```diff
@@ -784,11 +791,21 @@ static int rtadv_read(struct thread *thread)
+
+			zlog_debug("%s(%s) no RA instance found in VRF",
+				   __func__, VRF_LOGNAME(vrf));
+		}
```
is not it a bit too much verbose?

sworleys replied:
this was removed

pguibert6WIND reviewed (approved):
I understand this is a bit of optimisation.
is there a roadmap defined with RA to work on more than ptop bgp unnumbered ?

sworleys commented:
@pguibert6WIND don't merge this yet, I have some more changes I am gonna push to this PR.

sworleys commented:
> I understand this is a bit of optimisation.
> is there a roadmap defined with RA to work on more than ptop bgp unnumbered ?

No roadmap as far as I am aware. Just found a bug and took the time to do some changes in the RA code.

sworleys commented:
I simplified the code quite a bit. Instead of using a refs, I just point to the default VRF for vrf-lite scenarios.

I also found some edge cases around the interface counters for keeping track of interfaces that have been turned on for RAs. I extended this code to use a sorted list of interfaces by name rather than counters. I added a show command for this.

sworleys commented:
poke @pguibert6WIND you can re-review and merge now if you are happy with it still

pguibert6WIND commented on zebra/rtadv.c:
```diff
@@ -1385,6 +1385,76 @@ void zebra_interface_radv_enable(ZAPI_HANDLER_ARGS)
+		vty_out(vty,
+			"%% VRF subcommand only applicable for netns-based vrfs.\n");
+		return CMD_WARNING;
+	}
```
this command should work for netns based commands too.

pguibert6WIND replied:
actually, it is more the other way.
I don't see why forcing the user to dump interface infromation on default vrf, whereas the user may be vrf oriented, and may expect that the interface is located behind a give vrf.

sworleys replied:
Reposting what I said in slack for discussion here:


```
that command works for netns

more so actually
the vrf subcommand ONLY works for vrf-netns
since with vrf netns it's all handled in separate namespaces you need that info.
with vrf-lite, it's all in the default vrf, so using "vrf" doesn't do anything

with vrf-lite is doesn't make sense though. RAs are not handled in separate VRFs
this is mostly a debugging command so we can verify the state of the rtadv data

I could make it hidden.
```

I would prefer to not delimit it by VRFs for vrf-lite because that is not reflected in the actual state of the tables this is outputting.

pguibert6WIND reviewed (changes requested):
what about extending show command to netns too ?
other than that, changes Look Good To me. changes are zebra focused.

pguibert6WIND reviewed (approved):
implem is ok finally, I think.
after all, we reflect what the system does.

sworleys commented:
@pguibert6WIND can we merge this?
